### DISCUSSION TITLE
Reference docs

### DISCUSSION THREAD

**Discussion by Colpocorto** [RomanKubiak/ctrlr#186] [Q&A]
Hi,

is there any CTRLR reference manual anywhere?  I can't find any full, comprehensive documentation covering all aspects of CTRLR.
Besides the 3 tutorials on the CTRLR home page, and tons of info scattered around the forum, where is the rest of the info?

Thank you in advance.

**Comment by dnaldoog** [comment]
You already found it!

That’s all there is, apart from a 
manual written by Goodweather, but maybe the best way to get started is to look at the “Demo” panels in the installation folder of CTRLR . As well as other developers’ panels of course.

It takes a long time to start to learn it all, but it is very rewarding.

**Reply by Colpocorto** [reply]
Thank you for your response.

I don't consider that digging into demo code is a proper way to learn (especially if a developer hasn't too much spare time).
What "demo" panel should I check to learn about the meaning of a given field?
How can I determine why an example code is working but mine is not if I can't understand the basics of some aspects?
Where is explained how to use the Lua console? Where is explained how can I trace a function and know whether it is being executed or not... and why?

In my humble opinion, CTRLR is a great product and it has a big potential, but not having proper documentation makes me not consider it a serious alternative to JUCE/C++.  

I will finish the panel I'm working on but I will abandon the CTRLR path for further developments until it has decent documentation.

BR,

J.

**Reply by steev-m** [reply]
> That’s all there is, apart from a manual written by Goodweather

I've seen this mentioned numerous times, but the only links I can find are on the Ctrlr forums and they can't be clicked without having an account there. And you don't seem to be able to register any more. Is there a link to these docs anywhere?

**Reply by Ted-juh** [reply]
I might be wrong but I think they are referring to this manual:
[Ctrlr-Step-by-Step-v21.docx](https://github.com/RomanKubiak/ctrlr/files/10746693/Ctrlr-Step-by-Step-v21.docx)

**Reply by steev-m** [reply]
> I might be wrong but I think they are referring to this manual: [Ctrlr-Step-by-Step-v21.docx](https://github.com/RomanKubiak/ctrlr/files/10746693/Ctrlr-Step-by-Step-v21.docx)

Thanks a lot!

**Comment by Ted-juh** [comment]
You don't need a "demo" panel to find out the meaning of a given field.  Ctrlr itself can switch between Property ID's/Names and a description of what a given field is for by just ticking or unticking "Property ID's/ Names" under "View" in the Ctrlr Menubar. Refresh ID's/ Names and see what it does to the "General Properties". For instance: "modulatorValueExpression" changes to "Expression to evaluate when calculating the midi message value from the modulatorValue". 

To understand the basics, you have to keep a few things in mind. There's Juce, there's Lua, and there is CTRLR itself. Juce is well documented as is Lua. Ctrlr, I have to admit, is not that well documented. But you can find a lot of valuable information in the demo panels, in panels that other people made, at the old forums, and going over the open-source code of Ctrlr. Particularly the Source folder at the Ctrlr Github Code page. Oh, and search the old forums for a manual written by Goodweather. It's also a good starting point.

I never used the Lua console, never used the debugger. Comparator tables or Log viewer? They are not necessarily needed to build a panel. But I have to agree with you that CTRLR has quite a steep learning curve just because it's so versatile. There are many different ways to get the same result. It might not be the "best" way to do something but if the code or function works, it works. If it doesn't, throw it here in the discussions and some peeps here are very helpful to point you in the right direction or rewrite your code until it works.

For that matter. I tried Juce. My knowledge of C++ is limited. But I ran into the same problems I had with CTRLR. Half-baked tutorials, no explanation on why some code goes over here or over there. Serious, try to make a menubar. There is no tutorial for it and the demo file is so convoluted that it takes a while to find out what code is needed just for the menubar. And that is just because the "MenuBarModel Class" isn't that helpful either on how the MenuBar is constructed. Again, might be easy peasy for someone that knows his C++ but not for me.

Two things Roman got totally right. First, the WYSIWYG version of Juce that Ctrlr is. No continuous compiling like with Juce to see if the code works or if the Gui looks like what I want it to look like. And two the implementation of Lua. And that's just because Lua is far easier to learn than C++ in my opinion. That makes Ctrlr a very good alternative to Juce. And yes, it wouldn't hurt for CTRLR to have some sort of documentation but I rather see Roman working on CTRLR than writing documentation. 

But hey, that's just my opinion.

**Reply by Colpocorto** [reply]
I have written two VST synths with JUCE and never ran into documentation-related issues.

But don't get me wrong; I'm aware of the huge efforts that the developer has put into CTRLR and it seems indeed a great tool.

Sadly, my available time is limited and can't go with a tool that takes more time looking for info that actually developing software with it.
